src/Aspire.Dashboard/DashboardWebApplication.cs

```diff
@@ -212,6 +202,82 @@ public DashboardWebApplication(Action<WebApplicationBuilder>? configureBuilder =
         _app.MapGrpcService<OtlpMetricsService>();
         _app.MapGrpcService<OtlpTraceService>();
         _app.MapGrpcService<OtlpLogsService>();
+    }
+
+    // Kestrel endpoints are loaded from configuration. This is done so that advanced configuration of endpoints is
+    // possible from the caller. e.g., using environment variables to configure each endpoint's TLS certificate.
+    private void ConfigureKestrelEndpoints(WebApplicationBuilder builder, DashboardStartupConfiguration dashboardStartupConfig)
+    {
+        // Translate high-level config settings such as DOTNET_DASHBOARD_OTLP_ENDPOINT_URL and ASPNETCORE_URLS
+        // to Kestrel's schema for loading endpoints from configuration.
+        var browserEndpointNames = new List<string>(capacity: dashboardStartupConfig.BrowserUris.Length);
+        var initialValues = new Dictionary<string, string?>();
+        AddEndpointConfiguration(initialValues, "Otlp", dashboardStartupConfig.OtlpUri.OriginalString, HttpProtocols.Http2, requiredClientCertificate: dashboardStartupConfig.OtlpAuthMode == OtlpAuthMode.ClientCertificate);
+        if (dashboardStartupConfig.BrowserUris.Length == 1)
+        {
+            browserEndpointNames.Add("Browser");
+            AddEndpointConfiguration(initialValues, "Browser", dashboardStartupConfig.BrowserUris[0].OriginalString);
+        }
+        else
+        {
+            for (var i = 0; i < dashboardStartupConfig.BrowserUris.Length; i++)
+            {
+                var name = $"Browser{i}";
+                browserEndpointNames.Add(name);
+                AddEndpointConfiguration(initialValues, name, dashboardStartupConfig.BrowserUris[i].OriginalString);
+            }
+        }
+
+        static void AddEndpointConfiguration(Dictionary<string, string?> values, string endpointName, string url, HttpProtocols? protocols = null, bool requiredClientCertificate = false)
+        {
+            values[$"Kestrel:Endpoints:{endpointName}:Url"] = url;
+
+            if (protocols != null)
+            {
+                values[$"Kestrel:Endpoints:{endpointName}:Protocols"] = protocols.ToString();
+            }
+
+            if (requiredClientCertificate)
+            {
+                values[$"Kestrel:Endpoints:{endpointName}:ClientCertificateMode"] = ClientCertificateMode.RequireCertificate.ToString();
+            }
+        }
```

Comment:
My only question (non blocking) was why do this instead of using the API to create named endpoints?

Reply:
> * Change dashboard endpoint configuration from code to configuration + code. Allows [external endpoint configuration](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel/endpoints?view=aspnetcore-8.0#configure-endpoints-in-appsettingsjson) when starting the dashboard. This gives people who want to launch the dashboard a lot more flexibility over endpoint configuration. For example, the TLS cert can be different per endpoint via configuration, `Kestrel:Endpoints:Otlp:Certification:...=xxx`.

Without this, we'd need to recreate a lot of configuration infrastructure ourselves, i.e. provide a way to configure the TLS cert for each endpoint from a cert store, or from a pdf, or from a cert+key, etc.

Reply:
I see